### PR TITLE
feat: Refactors API client into hand rolled sdk in api/ directory 

### PR DESCRIPTION
**Pull request by jpmcb**
## Description

Refactors API client code into a hand rolled SDK under `api/`. This is partially for simplicity but also so we can move fast without having to roll out a new generated Go SDK.

This also refactors several of the broken uses of the `v2` API. Now, the old `pizza insights` commands work.

## Related Tickets & Documents

Closes: #87 

## Steps to QA

1. `just build`
2. `just lint`
3. `just test`
4. Able to run all the following commands:
    - `./build/pizza generate codeowners .`
    - `./build/pizza insights user-contributions`
    - `./build/pizza insights repositories`
    - `./build/pizza insights contributors`

## Tier (staff will fill in)

- [ ] Tier 1
- [ ] Tier 2
- [ ] Tier 3
- [x] Tier 4

### CODE REVIEW

**Review by brandonroberts — APPROVED** [review]

**Review by nickytonline — COMMENTED** [review]
I get the following error when running `just lint`. I also made sure I have all the deps, by running `go mod download`. It seems to just be the linting.

```
WARN [runner] Can't run linter goanalysis_metalinter: buildir: failed to load package : could not load export data: no export data for "github.com/cpuguy83/go-md2man/v2/md2man" 
INFO [runner] processing took 1.877µs with stages: max_same_issues: 334ns, invalid_issue: 250ns, source_code: 167ns, filename_unadjuster: 167ns, cgo: 167ns, sort_results: 125ns, identifier_marker: 83ns, path_prefixer: 42ns, skip_dirs: 42ns, path_shortener: 42ns, skip_files: 42ns, fixer: 42ns, nolint: 42ns, path_prettifier: 42ns, exclude: 42ns, max_from_linter: 42ns, severity-rules: 42ns, max_per_file_from_linter: 41ns, autogenerated_exclude: 41ns, exclude-rules: 41ns, diff: 41ns, uniq_by_line: 0s 
INFO [runner] linters took 1.844725876s with stages: goanalysis_metalinter: 1.844699918s 
ERRO Running error: can't run linter goanalysis_metalinter
buildir: failed to load package : could not load export data: no export data for "github.com/cpuguy83/go-md2man/v2/md2man" 
INFO Memory: 66 samples, avg is 137.4MB, max is 518.7MB 
INFO Execution took 6.436220378s                  
error: Recipe `lint` failed on line 159 with exit code 3
```

**Comment by nickytonline** [comment]
> I get the following error when running `just lint`. I also made sure I have all the deps, by running `go mod download`. It seems to just be the linting.
> 
> ```
> WARN [runner] Can't run linter goanalysis_metalinter: buildir: failed to load package : could not load export data: no export data for "github.com/cpuguy83/go-md2man/v2/md2man" 
> INFO [runner] processing took 1.877µs with stages: max_same_issues: 334ns, invalid_issue: 250ns, source_code: 167ns, filename_unadjuster: 167ns, cgo: 167ns, sort_results: 125ns, identifier_marker: 83ns, path_prefixer: 42ns, skip_dirs: 42ns, path_shortener: 42ns, skip_files: 42ns, fixer: 42ns, nolint: 42ns, path_prettifier: 42ns, exclude: 42ns, max_from_linter: 42ns, severity-rules: 42ns, max_per_file_from_linter: 41ns, autogenerated_exclude: 41ns, exclude-rules: 41ns, diff: 41ns, uniq_by_line: 0s 
> INFO [runner] linters took 1.844725876s with stages: goanalysis_metalinter: 1.844699918s 
> ERRO Running error: can't run linter goanalysis_metalinter
> buildir: failed to load package : could not load export data: no export data for "github.com/cpuguy83/go-md2man/v2/md2man" 
> INFO Memory: 66 samples, avg is 137.4MB, max is 518.7MB 
> INFO Execution took 6.436220378s                  
> error: Recipe `lint` failed on line 159 with exit code 3
> ```

It looks like running `go install github.com/golangci/golangci-lint/cmd/golangci-lint@latest` fixed the issue. I ran `go mod tidy` after that, but there doesn't appear to be any changes.

File: cmd/insights/repositories.go [hunk 1]
```diff
@@ -1,36 +1,35 @@
...
+	APIClient *api.Client
 
 	// Repos is the array of git repository urls
 	Repos []string
 
 	// FilePath is the path to yaml file containing an array of git repository urls
 	FilePath string
 
-	// Period is the number of days, used for query filtering
+	// RangeVal is the number of days, used for query filtering
```
Comment (nickytonline) [inline cmd/insights/repositories.go:30]: Is `Range` reserved or is `SomeNameVal` the naming convention for certain things in Go?

**Review by nickytonline — COMMENTED** [review]
Tested all the commands and they ran successfully. Just some questions.

CleanShot 2024-09-03 at 17 25 52

**Comment by jpmcb** [comment]
> Are linter tools in go land expected to be only on your local and not in the project?

No. There's one unified toolchain in the `go` command line. `golangci-lint` is an extra piece of sugar and has no integration into the Go toolchain. Think of it as an extra, optional opinion: as long as the `go build` or `go run` commands work, technically, it's completely valid Go.

File: cmd/insights/repositories.go (hunk 1 above)
Comment (jpmcb) [reply cmd/insights/repositories.go:30]: Yes, "range" is reserved. It's a keyword for iterating slices / maps / etc.:

```go
    sum := 0
    for _, num := range []int{2, 3, 4} {
        sum += num
    }
    fmt.Println("sum:", sum)
```

https://go.dev/play/p/_wuEx2nR4VS

https://gobyexample.com/range

**Comment by jpmcb** [comment]
> It looks like running go install github.com/golangci/golangci-lint/cmd/golangci-lint@latest 

That's very, very odd: the `just lint` command runs a container from their registry tagged `v1.59`:

https://github.com/open-sauced/pizza-cli/blob/ad9e3a0dd2ade42dbfdd1f6010af46eda35efb30/justfile#L152-L159

The local code directory is mounted as `/app` and the linting is run in a totally isolated environment. Do you consistently get that error? Even off `beta` or `main`?

You might also try to upgrade the linter locally. Looks like they cut a 1.60:

```diff
diff --git a/justfile b/justfile
index a0e24e0..f47a1ac 100644
--- a/justfile
+++ b/justfile
@@ -155,7 +155,7 @@ lint:
     --rm \
     -v "$(pwd)/:/app" \
     -w /app \
-    golangci/golangci-lint:v1.59 \
+    golangci/golangci-lint:v1.60 \
     golangci-lint run -v

 # Formats code via builtin go fmt
```

**Comment by nickytonline** [comment]
> > It looks like running go install github.com/golangci/golangci-lint/cmd/golangci-lint@latest
> 
> That's very, very odd: the `just lint` command runs a container from their registry tagged `v1.59`:
> 
> https://github.com/open-sauced/pizza-cli/blob/ad9e3a0dd2ade42dbfdd1f6010af46eda35efb30/justfile#L152-L159
> 
> Also note the successful lint run on this PR - maybe some useful logs there: [open-sauced/pizza-cli/actions/runs/10687979965/job/29626781164?pr=111](https://github.com/open-sauced/pizza-cli/actions/runs/10687979965/job/29626781164?pr=111)
> 
> The local code directory is mounted as `/app` and the linting is run in a totally isolated environment. Do you consistently get that error? Even off `beta` or `main`?
> 
> You might also try to upgrade the linter locally. Looks like they cut a 1.60:
> 
> ```diff
> diff --git a/justfile b/justfile
> index a0e24e0..f47a1ac 100644
> --- a/justfile
> +++ b/justfile
> @@ -155,7 +155,7 @@ lint:
>      --rm \
>      -v "$(pwd)/:/app" \
>      -w /app \
> -    golangci/golangci-lint:v1.59 \
> +    golangci/golangci-lint:v1.60 \
>      golangci-lint run -v
> 
>  # Formats code via builtin go fmt
> ```

My bad. Docker for Desktop wasn't completely running as I don't have it on by default. All good now.

```
pizza-cli on  api-client-refactor [$] via 🐳 desktop-linux via 🐹 v1.23.0 
❯ just lint
docker run -t --rm -v "$(pwd)/:/app" -w /app golangci/golangci-lint:v1.59 golangci-lint run -v
INFO golangci-lint has version 1.59.1 built with go1.22.3 from 1a55854a on 2024-06-09T18:08:33Z 
INFO [config_reader] Config search paths: [./ /app / /root] 
INFO [config_reader] Used config file .golangci.yaml 
INFO [lintersdb] Active 11 linters: [errcheck gci goimports gosimple govet ineffassign misspell revive staticcheck unconvert unused] 
INFO [loader] Go packages loading at mode 575 (deps|exports_file|types_sizes|name|compiled_files|files|imports) took 5.275550502s 
INFO [runner/filename_unadjuster] Pre-built 0 adjustments in 3.300542ms 
INFO [linters_context/goanalysis] analyzers took 10.827503202s with top 10 stages: buildir: 6.74468295s, the_only_name: 1.403188459s, inspect: 420.207254ms, fact_deprecated: 349.529408ms, printf: 277.186624ms, ctrlflow: 274.731882ms, nilness: 153.976679ms, fact_purity: 132.305999ms, gci: 113.724707ms, SA5012: 112.346914ms 
INFO [runner] Issues before processing: 57, after processing: 0 
INFO [runner] Processors filtering stat (out/in): skip_dirs: 57/57, exclude: 57/57, cgo: 57/57, filename_unadjuster: 57/57, invalid_issue: 57/57, identifier_marker: 57/57, skip_files: 57/57, autogenerated_exclude: 57/57, exclude-rules: 4/57, nolint: 0/4, path_prettifier: 57/57 
INFO [runner] processing took 4.082544ms with stages: autogenerated_exclude: 2.373792ms, exclude-rules: 740.666µs, identifier_marker: 555.458µs, nolint: 196.875µs, path_prettifier: 143.209µs, skip_dirs: 62.958µs, cgo: 2.792µs, invalid_issue: 2.125µs, filename_unadjuster: 1.958µs, max_same_issues: 667ns, fixer: 334ns, sort_results: 291ns, exclude: 209ns, skip_files: 208ns, uniq_by_line: 208ns, diff: 167ns, max_from_linter: 167ns, severity-rules: 126ns, max_per_file_from_linter: 125ns, path_shortener: 84ns, source_code: 83ns, path_prefixer: 42ns 
INFO [runner] linters took 1.825440167s with stages: goanalysis_metalinter: 1.821289542s 
INFO File cache stats: 45 entries of total size 98.8KiB 
INFO Memory: 73 samples, avg is 143.8MB, max is 592.3MB 
INFO Execution took 7.105187379s          
```

**Review by nickytonline — APPROVED** [review]
The CODEOWNERS issue isn't related to this so going to go ahead and ✅

**Pull request merged**